File: clients/typescript/test/client.test.ts
import assert from "node:assert/strict";
import { after, test } from "node:test";

import { start, type Serving } from "./harness.js";
import {
  bool,
  bytes,
  float,
  ge,
  int,
  isKind,
  nullValue,
  SlateError,
  str,
  uint,
  uuid,
  type Value,
} from "../src/index.js";

const servers: Serving[] = [];

async function serving(extra = ""): Promise<Serving> {
  const s = await start(extra);
  servers.push(s);
  return s;
}

after(() => {
  for (const s of servers) s.stop();
});

function expectUint(value: Value | undefined, want: bigint, what: string): void {
  assert.ok(value, `${what} is missing`);
  assert.equal(value.kind, "uint", `${what} should be a uint`);
  assert.equal((value as { value: bigint }).value, want, what);
}

test("insert then get", async () => {
  const session = (await serving()).client().session();

  const result = await session.insert("docs", [uint(1), str("note"), int(10)]);
  assert.equal(result.affected, 1n);
  // The writer's position, which is what a caller carries to another session
  // to read its own write there.
  assert.ok(result.sequence !== undefined, "a write should report its sequence");

  const row = await session.get("docs", [uint(1)]);
  assert.ok(row, "the row just inserted was not found");
  assert.equal(row[1]?.kind, "string");
  assert.equal((row[1] as { value: string }).value, "note");
  assert.equal((row[2] as { value: bigint }).value, 10n);
});

// A missing row is `undefined`, not a throw: checking existence should not
// mean catching an exception.
test("a missing row is undefined rather than an error", async () => {
  const session = (await serving()).client().session();
  assert.equal(await session.get("docs", [uint(404)]), undefined);
});

test("insert refuses a duplicate key", async () => {
  const session = (await serving()).client().session();
  const row = [uint(1), str("a"), int(1)];
  await session.insert("docs", row);

  await assert.rejects(
    () => session.insert("docs", row),
    (error: unknown) => {
      assert.ok(isKind(error, "already-exists"), `kind was ${(error as SlateError).kind}`);
      assert.equal((error as SlateError).retryable, false, "a taken key is not retryable");
      return true;
    },
  );
});

// Upsert is the same call with a flag, so it is worth pinning that the flag
// reaches the server rather than the two behaving identically.
test("upsert replaces where insert refuses", async () => {
  const session = (await serving()).client().session();
  await session.insert("docs", [uint(1), str("first"), int(1)]);
  await session.upsert("docs", [uint(1), str("second"), int(2)]);

  const row = await session.get("docs", [uint(1)]);
  assert.equal((row?.[1] as { value: string }).value, "second");
});

test("query filters and orders", async () => {
  const session = (await serving()).client().session();
  for (let i = 1n; i <= 5n; i++) {
    await session.insert("docs", [uint(i), str("k"), int(i * 10n)]);
  }

  const rows = await session
    .query({
      table: "docs",
      filter: ge(2, int(30)),
      sort: [{ column: 2, direction: "desc" }],
    })
    .collect();

  assert.equal(rows.length, 3, "sizes 30, 40 and 50 match");
  assert.equal((rows[0]?.[2] as { value: bigint }).value, 50n);
  assert.equal((rows[2]?.[2] as { value: bigint }).value, 30n);
});

test("limit and projection", async () => {
  const session = (await serving()).client().session();
  for (let i = 1n; i <= 4n; i++) {
    await session.insert("docs", [uint(i), str("k"), int(i)]);
  }

  const rows = await session
    .query({ table: "docs", limit: 2, columns: [0] })
    .collect();

  assert.equal(rows.length, 2);
  // Columns outside the projection come back null rather than absent, so the
  // row keeps its shape and an ordinal still means what it meant.
  assert.equal(rows[0]?.[1]?.kind, "null");
});

test("a rolled-back transaction discards its writes", async () => {
  const session = (await serving()).client().session();
  const tx = await session.begin();

  await tx.insert("docs", [uint(7), str("x"), int(1)]);
  assert.ok(await tx.get("docs", [uint(7)]), "a transaction must see its own writes");

  await tx.rollback();
  assert.equal(await session.get("docs", [uint(7)]), undefined);
});

test("a committed transaction lands", async () => {
  const session = (await serving()).client().session();
  const tx = await session.begin();
  await tx.insert("docs", [uint(8), str("y"), int(2)]);
  await tx.commit();

  assert.ok(await session.get("docs", [uint(8)]));
});

// `finally { await tx.rollback() }` is the shape every transaction should
// use, so rolling back after a commit has to be quiet.
test("rollback after commit is quiet", async () => {
  const session = (await serving()).client().session();
  const tx = await session.begin();
  await tx.commit();
  await tx.rollback();
});

test("delete", async () => {
  const session = (await serving()).client().session();
  await session.insert("docs", [uint(3), str("z"), int(1)]);

  const result = await session.delete("docs", [uint(3)]);
  assert.equal(result.affected, 1n);
  assert.equal(await session.get("docs", [uint(3)]), undefined);
});

// Every value kind must survive the round trip. A client that encodes a u64 as
// an i64 gets a different value back from this server, because its ordering is
// type-first, so this is not a formality.
test("every value kind round-trips", async () => {
  const server = await serving(`
[[tables]]
name = "kinds"
id = 2
columns = [
  { name = "id",  type = "u64" },
  { name = "b",   type = "bool" },
  { name = "by",  type = "bytes" },
  { name = "s",   type = "str" },
  { name = "i",   type = "i64" },
  { name = "u",   type = "u64" },
  { name = "f",   type = "f64" },
  { name = "uu",  type = "uuid" },
]
primary_key = ["id"]

[[security.grants]]
role = "app"
tables = ["kinds"]
actions = ["everything"]
`);
  const session = server.client().session();
  const id = new Uint8Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]);

  await session.insert("kinds", [
    uint(1),
    bool(true),
    bytes(new Uint8Array([0xde, 0xad])),
    str("hello"),
    int(-5),
    uint(5),
    float(1.5),
    uuid(id),
  ]);

  const row = await session.get("kinds", [uint(1)]);
  assert.ok(row);
  assert.equal((row[1] as { value: boolean }).value, true);
  assert.deepEqual(Array.from((row[2] as { value: Uint8Array }).value), [0xde, 0xad]);
  assert.equal((row[3] as { value: string }).value, "hello");
  assert.equal((row[4] as { value: bigint }).value, -5n);
  expectUint(row[5], 5n, "the u64 column");
  assert.equal((row[6] as { value: number }).value, 1.5);
  assert.deepEqual(Array.from((row[7] as { value: Uint8Array }).value), Array.from(id));

  // An i64 and a u64 must not decode to the same value even at the same
  // magnitude — this server orders type-first.
  assert.notEqual(row[4]?.kind, row[5]?.kind);
});

// A u64 above 2^53 must survive, which is the whole reason values are bigint
// rather than number. A `number` would lose the low bits silently, and a
// primary key is where that shows up latest.
test("a u64 past the double's precision survives", async () => {
  const session = (await serving()).client().session();
  const big = 9_007_199_254_740_993n; // 2^53 + 1

  await session.insert("docs", [uint(big), str("big"), int(1)]);
  const row = await session.get("docs", [uint(big)]);
  assert.ok(row, "a large key must be findable again");
  expectUint(row[0], big, "the large primary key");
});

test("explain needs the explain grant", async () => {
  const server = await serving(`
[[tables]]
name = "plain"
id = 3
columns = [{ name = "id", type = "u64" }]
primary_key = ["id"]

[[security.grants]]
role = "reader"
tables = ["plain"]
actions = ["all"]
`);
  const session = server
    .client({ principal: "u64:2", tenant: "u64:1", roles: ["reader"] })
    .session();

  await assert.rejects(
    () => session.explain({ table: "plain" }),
    (error: unknown) => {
      assert.ok(
        isKind(error, "permission-denied"),
        `a read grant must not carry EXPLAIN, got ${(error as SlateError).kind}`,
      );
      return true;
    },
  );
});

test("explain with the grant", async () => {
  const session = (await serving()).client().session();
  await session.insert("docs", [uint(1), str("k"), int(1)]);

  const plan = await session.explain({ table: "docs" });
  assert.equal(plan.table, "docs");
  assert.notEqual(plan.access, "", "a plan should name its access path");
});

test("leadership", async () => {
  const status = await (await serving()).client().leadership();
  assert.equal(status.leader, true, "a lone node should hold the lease");
});

// A session's watermark must advance past its own writes, which is what makes
// a later read see them.
test("the session watermark advances on write", async () => {
  const session = (await serving()).client().session();
  assert.equal(session.watermark, undefined, "a fresh session has no watermark");

  await session.insert("docs", [uint(1), str("k"), int(1)]);
  const first = session.watermark;
  assert.ok(first !== undefined && first > 0n, "a write must advance the watermark");

  await session.insert("docs", [uint(2), str("k"), int(2)]);
  const second = session.watermark;
  assert.ok(
    second !== undefined && second > first,
    "a second write must advance it further, not merely set it once",
  );
});

test("an unknown table is not-found", async () => {
  const session = (await serving()).client().session();
  await assert.rejects(
    () => session.get("nosuchtable", [uint(1)]),
    (error: unknown) => {
      assert.ok(isKind(error, "not-found"), `kind was ${(error as SlateError).kind}`);
      return true;
    },
  );
});

// A null is not a zero and not an absent column, and the round trip has to
// keep all three apart. `docs.kind` is not nullable — the server refuses one
// there, correctly — so this needs a column that is.
test("a null value is not a zero", async () => {
  const server = await serving(`
[[tables]]
name = "maybe"
id = 4
columns = [
  { name = "id",   type = "u64" },
  { name = "note", type = "str", nullable = true },
  { name = "n",    type = "i64" },
]
primary_key = ["id"]

[[security.grants]]
role = "app"
tables = ["maybe"]
actions = ["everything"]
`);
  const session = server.client().session();
  await session.insert("maybe", [uint(1), nullValue, int(0)]);

  const row = await session.get("maybe", [uint(1)]);
  assert.equal(row?.[1]?.kind, "null", "a written null must come back null");
  assert.equal(row?.[2]?.kind, "int", "and a zero must stay an int");
  assert.equal((row?.[2] as { value: bigint }).value, 0n);
});

// The server refuses a null in a column that is not nullable, and the client
// must surface that as an ordinary refusal rather than swallowing it.
test("a null in a non-nullable column is refused", async () => {
  const session = (await serving()).client().session();
  await assert.rejects(
    () => session.insert("docs", [uint(1), nullValue, int(0)]),
    (error: unknown) => {
      assert.ok(isKind(error, "invalid-request"), `kind was ${(error as SlateError).kind}`);
      assert.match((error as SlateError).message, /nullable/);
      return true;
    },
  );
});

File: clients/go/slate/client_test.go
package slate_test

import (
	"errors"
	"testing"

	"github.com/howlerops/slate-orm/clients/go/slate"
)

func rows(t *testing.T, s *slate.RowStream) [][]slate.Value {
	t.Helper()
	out, err := s.Collect()
	if err != nil {
		t.Fatalf("draining the stream: %v", err)
	}
	return out
}

func TestInsertThenGet(t *testing.T) {
	server := start(t, "")
	session := server.client(t).Session()
	ctx := testContext(t)

	result, err := session.Insert(ctx, "docs",
		[]slate.Value{slate.Uint(1), slate.String("note"), slate.Int(10)})
	if err != nil {
		t.Fatalf("inserting: %v", err)
	}
	if result.Affected != 1 {
		t.Errorf("affected = %d, want 1", result.Affected)
	}
	// The writer's position after the write, which is what a caller carries to
	// another session to read its own write there.
	if result.Sequence == nil {
		t.Error("a write should report the sequence it landed at")
	}

	row, found, err := session.Get(ctx, "docs", []slate.Value{slate.Uint(1)})
	if err != nil {
		t.Fatalf("getting: %v", err)
	}
	if !found {
		t.Fatal("the row just inserted was not found")
	}
	if len(row) != 3 || row[1] != slate.String("note") || row[2] != slate.Int(10) {
		t.Errorf("row = %v, want [1 note 10]", row)
	}
}

// A missing row is `found == false`, not an error: checking existence should
// not mean matching on an error type.
func TestGetMissingIsNotAnError(t *testing.T) {
	server := start(t, "")
	session := server.client(t).Session()
	ctx := testContext(t)

	row, found, err := session.Get(ctx, "docs", []slate.Value{slate.Uint(404)})
	if err != nil {
		t.Fatalf("getting a missing row should not error: %v", err)
	}
	if found {
		t.Errorf("found a row that was never written: %v", row)
	}
}

func TestInsertRefusesADuplicateKey(t *testing.T) {
	server := start(t, "")
	session := server.client(t).Session()
	ctx := testContext(t)

	row := []slate.Value{slate.Uint(1), slate.String("a"), slate.Int(1)}
	if _, err := session.Insert(ctx, "docs", row); err != nil {
		t.Fatalf("first insert: %v", err)
	}
	_, err := session.Insert(ctx, "docs", row)
	if err == nil {
		t.Fatal("a duplicate primary key must be refused")
	}
	if !slate.IsKind(err, slate.KindAlreadyExists) {
		t.Errorf("err = %v, want KindAlreadyExists", err)
	}
	var e *slate.Error
	if errors.As(err, &e) && e.Retryable() {
		t.Error("a duplicate key is not retryable")
	}
}

// Upsert is the same call with a flag, so it is worth pinning that the flag
// actually reaches the server rather than the two behaving identically.
func TestUpsertReplacesWhereInsertRefuses(t *testing.T) {
	server := start(t, "")
	session := server.client(t).Session()
	ctx := testContext(t)

	if _, err := session.Insert(ctx, "docs",
		[]slate.Value{slate.Uint(1), slate.String("first"), slate.Int(1)}); err != nil {
		t.Fatalf("inserting: %v", err)
	}
	if _, err := session.Upsert(ctx, "docs",
		[]slate.Value{slate.Uint(1), slate.String("second"), slate.Int(2)}); err != nil {
		t.Fatalf("upserting: %v", err)
	}
	row, found, err := session.Get(ctx, "docs", []slate.Value{slate.Uint(1)})
	if err != nil || !found {
		t.Fatalf("get after upsert: %v found=%v", err, found)
	}
	if row[1] != slate.String("second") {
		t.Errorf("row = %v, want the upserted values", row)
	}
}

func TestQueryFiltersAndOrders(t *testing.T) {
	server := start(t, "")
	session := server.client(t).Session()
	ctx := testContext(t)

	for i := uint64(1); i <= 5; i++ {
		if _, err := session.Insert(ctx, "docs", []slate.Value{
			slate.Uint(i), slate.String("k"), slate.Int(int64(i) * 10),
		}); err != nil {
			t.Fatalf("seeding %d: %v", i, err)
		}
	}

	stream, err := session.Query(ctx, slate.Query{
		Table:  "docs",
		Filter: slate.Filter(slate.Ge(2, slate.Int(30))),
		Sort:   []slate.SortKey{{Column: 2, Direction: slate.Desc}},
	})
	if err != nil {
		t.Fatalf("querying: %v", err)
	}
	got := rows(t, stream)
	if len(got) != 3 {
		t.Fatalf("got %d rows, want 3 (sizes 30, 40, 50)", len(got))
	}
	if got[0][2] != slate.Int(50) || got[2][2] != slate.Int(30) {
		t.Errorf("rows are not in descending size order: %v", got)
	}
}

func TestQueryLimitAndProjection(t *testing.T) {
	server := start(t, "")
	session := server.client(t).Session()
	ctx := testContext(t)

	for i := uint64(1); i <= 4; i++ {
		if _, err := session.Insert(ctx, "docs", []slate.Value{
			slate.Uint(i), slate.String("k"), slate.Int(int64(i)),
		}); err != nil {
			t.Fatalf("seeding: %v", err)
		}
	}

	stream, err := session.Query(ctx, slate.Query{
		Table:   "docs",
		Limit:   slate.Limit(2),
		Columns: []slate.Ordinal{0},
	})
	if err != nil {
		t.Fatalf("querying: %v", err)
	}
	got := rows(t, stream)
	if len(got) != 2 {
		t.Fatalf("got %d rows, want 2", len(got))
	}
	// Columns outside the projection come back null rather than absent, so the
	// row keeps its shape and an ordinal still means what it meant.
	if _, isNull := got[0][1].(slate.Null); !isNull {
		t.Errorf("an unprojected column should be null, got %v", got[0][1])
	}
}

func TestTransactionRollbackDiscards(t *testing.T) {
	server := start(t, "")
	session := server.client(t).Session()
	ctx := testContext(t)

	tx, err := session.Begin(ctx)
	if err != nil {
		t.Fatalf("beginning: %v", err)
	}
	if _, err := tx.Insert(ctx, "docs",
		[]slate.Value{slate.Uint(7), slate.String("x"), slate.Int(1)}); err != nil {
		t.Fatalf("inserting in the transaction: %v", err)
	}
	// Visible inside, before the commit.
	if _, found, err := tx.Get(ctx, "docs", []slate.Value{slate.Uint(7)}); err != nil || !found {
		t.Fatalf("a transaction must see its own writes: %v found=%v", err, found)
	}
	if err := tx.Rollback(ctx); err != nil {
		t.Fatalf("rolling back: %v", err)
	}
	if _, found, err := session.Get(ctx, "docs", []slate.Value{slate.Uint(7)}); err != nil || found {
		t.Fatalf("a rolled-back write must not be visible: %v found=%v", err, found)
	}
}

func TestTransactionCommitLands(t *testing.T) {
	server := start(t, "")
	session := server.client(t).Session()
	ctx := testContext(t)

	tx, err := session.Begin(ctx)
	if err != nil {
		t.Fatalf("beginning: %v", err)
	}
	defer func() { _ = tx.Rollback(ctx) }()

	if _, err := tx.Insert(ctx, "docs",
		[]slate.Value{slate.Uint(8), slate.String("y"), slate.Int(2)}); err != nil {
		t.Fatalf("inserting: %v", err)
	}
	if err := tx.Commit(ctx); err != nil {
		t.Fatalf("committing: %v", err)
	}
	if _, found, err := session.Get(ctx, "docs", []slate.Value{slate.Uint(8)}); err != nil || !found {
		t.Fatalf("a committed write must be visible: %v found=%v", err, found)
	}
}

// `defer tx.Rollback(ctx)` is the shape every transaction should use, so
// rolling back after a commit has to be a no-op rather than an error.
func TestRollbackAfterCommitIsQuiet(t *testing.T) {
	server := start(t, "")
	session := server.client(t).Session()
	ctx := testContext(t)

	tx, err := session.Begin(ctx)
	if err != nil {
		t.Fatalf("beginning: %v", err)
	}
	if err := tx.Commit(ctx); err != nil {
		t.Fatalf("committing: %v", err)
	}
	if err := tx.Rollback(ctx); err != nil {
		t.Errorf("rollback after commit should be quiet, got %v", err)
	}
}

func TestDelete(t *testing.T) {
	server := start(t, "")
	session := server.client(t).Session()
	ctx := testContext(t)

	if _, err := session.Insert(ctx, "docs",
		[]slate.Value{slate.Uint(3), slate.String("z"), slate.Int(1)}); err != nil {
		t.Fatalf("inserting: %v", err)
	}
	result, err := session.Delete(ctx, "docs", []slate.Value{slate.Uint(3)})
	if err != nil {
		t.Fatalf("deleting: %v", err)
	}
	if result.Affected != 1 {
		t.Errorf("affected = %d, want 1", result.Affected)
	}
	if _, found, _ := session.Get(ctx, "docs", []slate.Value{slate.Uint(3)}); found {
		t.Error("the deleted row is still there")
	}
}

// Every value kind must survive the round trip. A client that encodes a u64 as
// an i64 gets a different value back from this server, because its ordering is
// type-first — so this is not a formality.
func TestEveryValueKindRoundTrips(t *testing.T) {
	server := start(t, `
[[tables]]
name = "kinds"
id = 2
columns = [
  { name = "id",  type = "u64" },
  { name = "b",   type = "bool" },
  { name = "by",  type = "bytes" },
  { name = "s",   type = "str" },
  { name = "i",   type = "i64" },
  { name = "u",   type = "u64" },
  { name = "f",   type = "f64" },
  { name = "uu",  type = "uuid" },
]
primary_key = ["id"]

[[security.grants]]
role = "app"
tables = ["kinds"]
actions = ["everything"]
`)
	session := server.client(t).Session()
	ctx := testContext(t)

	uu := slate.UUID{0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
		0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10}
	written := []slate.Value{
		slate.Uint(1), slate.Bool(true), slate.Bytes{0xde, 0xad},
		slate.String("hello"), slate.Int(-5), slate.Uint(5),
		slate.Float(1.5), uu,
	}
	if _, err := session.Insert(ctx, "kinds", written); err != nil {
		t.Fatalf("inserting: %v", err)
	}

	read, found, err := session.Get(ctx, "kinds", []slate.Value{slate.Uint(1)})
	if err != nil || !found {
		t.Fatalf("getting: %v found=%v", err, found)
	}
	if read[1] != slate.Bool(true) {
		t.Errorf("bool = %v", read[1])
	}
	if string(read[2].(slate.Bytes)) != "\xde\xad" {
		t.Errorf("bytes = %v", read[2])
	}
	if read[3] != slate.String("hello") {
		t.Errorf("string = %v", read[3])
	}
	if read[4] != slate.Int(-5) {
		t.Errorf("int = %v", read[4])
	}
	if read[5] != slate.Uint(5) {
		t.Errorf("uint = %v", read[5])
	}
	if read[6] != slate.Float(1.5) {
		t.Errorf("float = %v", read[6])
	}
	if read[7] != uu {
		t.Errorf("uuid = %v, want %v", read[7], uu)
	}
	// An i64 and a u64 of the same magnitude must not come back the same.
	if read[4] == read[5] {
		t.Error("i64 -5 and u64 5 must not decode to the same value")
	}
}

func TestExplainNeedsTheExplainGrant(t *testing.T) {
	// This node grants `all` — the four data actions — and not `explain`.
	server := start(t, `
[[tables]]
name = "plain"
id = 3
columns = [{ name = "id", type = "u64" }]
primary_key = ["id"]

[[security.grants]]
role = "reader"
tables = ["plain"]
actions = ["all"]
`)
	c, err := slate.Dial(server.addr, slate.Identity{
		Principal: "u64:2", Tenant: "u64:1", Roles: []string{"reader"},
	})
	if err != nil {
		t.Fatalf("dialling: %v", err)
	}
	defer func() { _ = c.Close() }()
	ctx := testContext(t)

	_, err = c.Session().Explain(ctx, slate.Query{Table: "plain"})
	if err == nil {
		t.Fatal("a `read` grant must not carry EXPLAIN")
	}
	if !slate.IsKind(err, slate.KindPermissionDenied) {
		t.Errorf("err = %v, want KindPermissionDenied", err)
	}
}

func TestExplainWithTheGrant(t *testing.T) {
	server := start(t, "")
	session := server.client(t).Session()
	ctx := testContext(t)

	if _, err := session.Insert(ctx, "docs",
		[]slate.Value{slate.Uint(1), slate.String("k"), slate.Int(1)}); err != nil {
		t.Fatalf("seeding: %v", err)
	}
	plan, err := session.Explain(ctx, slate.Query{Table: "docs"})
	if err != nil {
		t.Fatalf("explaining: %v", err)
	}
	if plan.Table != "docs" {
		t.Errorf("plan is for %q, want docs", plan.Table)
	}
	if plan.Access == "" {
		t.Error("a plan should name its access path")
	}
}

func TestLeadership(t *testing.T) {
	server := start(t, "")
	ctx := testContext(t)

	status, err := server.client(t).Leadership(ctx)
	if err != nil {
		t.Fatalf("asking about leadership: %v", err)
	}
	if !status.Leader {
		t.Errorf("a lone node should hold the lease: %+v", status)
	}
}

// A session's watermark must advance past its own writes, which is what makes
// a later read see them.
func TestSessionWatermarkAdvancesOnWrite(t *testing.T) {
	server := start(t, "")
	session := server.client(t).Session()
	ctx := testContext(t)

	if before := session.Watermark(); before != nil {
		t.Errorf("a fresh session has no watermark, got %v", *before)
	}
	if _, err := session.Insert(ctx, "docs",
		[]slate.Value{slate.Uint(1), slate.String("k"), slate.Int(1)}); err != nil {
		t.Fatalf("inserting: %v", err)
	}
	after := session.Watermark()
	if after == nil {
		t.Fatal("a write must advance the session's watermark")
	}
	// A second write must move it further, not merely set it once.
	if _, err := session.Insert(ctx, "docs",
		[]slate.Value{slate.Uint(2), slate.String("k"), slate.Int(2)}); err != nil {
		t.Fatalf("second insert: %v", err)
	}
	later := session.Watermark()
	if later == nil || *later <= *after {
		t.Errorf("the watermark should advance again: %v then %v", after, later)
	}
	if *after == 0 {
		t.Error("the watermark should be a real sequence")
	}
}

func TestUnknownTableIsNotFound(t *testing.T) {
	server := start(t, "")
	session := server.client(t).Session()
	ctx := testContext(t)

	_, _, err := session.Get(ctx, "nosuchtable", []slate.Value{slate.Uint(1)})
	if err == nil {
		t.Fatal("an unknown table must be refused")
	}
	if !slate.IsKind(err, slate.KindNotFound) {
		t.Errorf("err = %v, want KindNotFound", err)
	}
}
